circleguard/investigator.py

```diff
@@ -226,43 +250,154 @@ def median_frametime(frametimes):
-                object_i += 1
-
-        return array
+                hitobj_type = 2
+                hitobj_end_time = hitobj.end_time.total_seconds() * 1000
+
+            # before sliderbug fix, notelock ended after hitwindow50
+            if version < version_sliderbug_fixed:
+                notelock_end_time = hitobj_t + hitwindow
+                # exception for sliders/spinners, where notelock ends after
+                # hitobject end time if it's earlier
+                if hitobj_type != 0:
+                    notelock_end_time = min(notelock_end_time, hitobj_end_time)
+            # after sliderbug fix, notelock ends after hitobject end time
+            else:
+                notelock_end_time = hitobj_end_time
+                # apparently notelock was increased by 2ms for circles
+                # (from testing)
+                if hitobj_type == 0:
+                    notelock_end_time += hitwindow + 2
+                # account for 0 frames that cause notelock to be 1ms shorter
+                if keydowns[keydown_i][2]:
+                    notelock_end_time -= 1
+
+
+            # can't press on hitobjects before hitwindowmiss
+            if keydown_t < hitobj_t - 399.5:
+                keydown_i += 1
+                continue
+
+            if keydown_t < hitobj_t - hitwindow:
+                # pressing on a circle or slider during hitwindowmiss will cause
+                #  a miss
+                if np.linalg.norm(keydown_xy - hitobj_xy) <= hitradius and hitobj_type != 2:
+
+                    # sliders don't disappear after missing
+                    # so we skip to the press_i that is after notelock_end_time
+                    keydown_i += 1
+                    if keydown_i < len(keydowns) and hitobj_type == 1 and version >= version_sliderbug_fixed:
+                        while keydowns[keydown_i][0] <= notelock_end_time:
+                            keydown_i += 1
+                            if keydown_i >= len(keydowns):
+                                break
+                    hitobj_i += 1
+                # keypress not on object, so we move to the next keypress
+                else:
+                    keydown_i += 1
+            elif keydown_t >= notelock_end_time:
+                # can no longer interact with hitobject after notelock_end_time
+                # so we move to the next object
+                hitobj_i += 1
+            else:
+                if keydown_t < hitobj_t + hitwindow and np.linalg.norm(keydown_xy - hitobj_xy) <= hitradius and hitobj_type != 2:
+                    hits.append([hitobj_t, keydown_t])
+
+                    # sliders don't disappear after clicking
+                    # so we skip to the press_i that is after notelock_end_time
+                    keydown_i += 1
+                    if keydown_i < len(keydowns) and hitobj_type == 1 and version >= version_sliderbug_fixed:
+                        while keydowns[keydown_i][0] <= notelock_end_time:
```

Comment:
you use `>= notelock_end_time` and ` <= notelock_end_time`, which is it? I don't think they can both be true; either a keydown at exactly the time notelock ends interacts with the note, or it doesn't

Reply:
thx for noticing that, completely missed it. I'm actually not sure which one it is. Will do some tests later, since I also need to figure out if the game rounds slider end time to fix this issue.